### PR TITLE
Implement disabling input through telon

### PR DESCRIPTION
**Pull request by mjtorn**
Use this for cutscenes which have any sort of timing
related to them, because user input like skipping dialog
would break it.

Disables saving as well, because loading a game from the
middle of a cutscene can't lead to anything good.

### CODE REVIEW

**Comment by fleskesvor** [comment]
There's already similar functionality in telon.gd. I'm not 100% sure what the state of it is, but it would be great to harmonise it with pre-existing logic. I'm thinking that if it isn't already possible to trigger from esc code, it might be more idiomatic to disable input via `set_state`, since telon is already registered as an object.

**Comment by mjtorn** [comment]
This is probably too hard to pull off In telon. The animation would essentially have to call Godot functions, which I don't think is possible.

Currently the telon code is only called when running an event.

Is there a risk that autosave gets hit in the middle of an event? I mean `global_vm.run_event()` disables input and starts running things off the stack; if something on the stack causes a save, loading that will put the game in a state which is in mid-stack. It's probably a very thin race condition, if it exists, but setting the global flag is guaranteed to work as a lock, right?

IMO an alternative solution would be to make `main.set_input_catch()` more Godot-idiomatic and disable input through the root node. Then we'd get rid of input-catching in telon altogether.

So basically keep this PR as it is but add a commit to clean up talon.

@StraToN are you the authority on this? What do you think?

**Comment by StraToN** [comment]
> This is probably too hard to pull off In telon. The animation would essentially have to call Godot functions, which I don't think is possible.

Yes it is, with the all-new AnimationPlayer node. You can add a "Call method track" in an animation, that call a node's method at the defined key(s). Is that what you need?

**Comment by mjtorn** [comment]
@StraToN that might do the trick, but what do we do while we wait?

Should we merge this and have it coexist with a fancier telon system later?

Should I just keep this PR open as a reminder and keep my code until 3.1 comes out?

Because until 3.1 is stable, I don't think this is a solvable problem using the telon.

**Comment by StraToN** [comment]
> Should we merge this and have it coexist with a fancier telon system later?

Seems like a good solution. Waiting for @fleskesvor 's opinion on this, but as long as we set a related issue, I'm fine with merging this PR before we come out with a better solution with Telon.

File: device/globals/global_vm.gd [hunk 1]
```diff
@@ -328,6 +328,10 @@ func run_event(p_event):
 func sched_event(time, obj, event):
 	event_queue.push_back([time, obj, event])
 
+func set_input_disabled(p_input_disabled):
+	$"/root".set_disable_input(p_input_disabled)
```
[the remark this answers is not in the document]
Comment (StraToN) [reply device/globals/global_vm.gd]: Yep, this method doesn't exist anymore in Godot 3.x (that was a Viewport function).
It has been replaced by `set_gui_disable_input(disabled).

**Comment by StraToN** [comment]
See comment https://github.com/godotengine/escoria/pull/203#discussion_r198881829

**Comment by StraToN** [comment]
We'll need a functionality present in 3.1. Labeling this PR as blocked until Godot 3.1 stable comes out.

**Comment by mjtorn** [comment]
It's agreed we'll leave this PR pending as a reminder, and I'll keep this in my fork, until 3.1 is out.

Then we'll use telon to deal with it. We must also remember to disable saving when input is blocked for the reasons given above.

**Comment by mjtorn** [comment]
Turned out the relevant functionality, calling functions in animations, was always there. I altered the functionality to match the telon interface.

I got nothing but bugs trying to use `cut_scene telon catch_input` and `cut_scene telon release_input`.  One of them is that clicks (when trying to walk) get the coordinates (0, 0). I believe this might have something to do with how `MOUSE_FILTER_*` interacts with `Control` nodes, but I don't understand the details. Something about how there is no `Control` node to handle the event?

This is why I removed the call to play the `release_input` animation once it was implemented, it caused nothing but misery and pain. Before this, it caused complaints. So removing it was not a regression.

What I really cared about with this PR is disabling input during cut scenes, so because I couldn't get the existing functionality to "play card", I implemented new animations. I understood telon is the preferred way of doing it anyway. The obvious benefit is that a game can "fork" telon and define the new one in `project.settings` and add actual visual animation tracks to when input is disabled, so that's fine.

This solution works in our jam game, by using `cut_scene telon disable_input` and its counterpart.

I propose merging this, but someone who understands telon and `catch_input` should take a further look into cleaning it up and documenting it.

Questions remaining:

  * Are the names `catch_input` and `release_input` descriptive or should they be renamed? To me "catch" sounds like input would be ignored, but in this context it's required to skip dialog
  * Is there a possibility that an autosave happens while running a `:load` event stack, corrupting the game state on loading this autosave?
  * Is something wrong when using `catch_input` and `release_input` when running other events? If `:load` needs to disable saving and input, `run_event()` should discern between that and a `:ready` event.

On the last question: It's really problematic to disable input for all events, because then a dialog tree in `:ready` cannot be played.

Anyway, I don't really understand telon or the interactions here; all I can say is this code works for me and breaks nothing I could find, though it might not be the final version for life.

**Comment by StraToN** [comment]
> Are the names catch_input and release_input descriptive or should they be renamed? To me "catch" sounds like input would be ignored, but in this context it's required to skip dialog

These names aren't indeed very descriptive. At least `catch_input` is OK for me since in the point of view of the telon itself, it defines whether to take the input or not. However, `release_input` is not understandable : which input: past input ? Next input? Not clear.

> Is there a possibility that an autosave happens while running a :load event stack, corrupting the game state on loading this autosave?

Autosaves should be simply forbidden while running a `:load` event stack. Maybe check this in global_vm while loading the ESC script to warn the user and simply ignore it?

> Is something wrong when using `catch_input` and `release_input` when running other events? If `:load` needs to disable saving and input, `run_event()` should discern between that and a `:ready` event.

I agree on this.

**Comment by mjtorn** [comment]
> At least catch_input is OK for me since in the point of view of the telon itself, it defines whether to take the input or not. However, release_input is not understandable : which input: past input ? Next input? Not clear.

Yeah, exactly, and telon catches input to skip dialog, but the input doesn't terminate there. The background gets input too. That's something I didn't understand at first. But what should `release_input` be called?

> Autosaves should be simply forbidden while running a :load event stack. Maybe check this in global_vm while loading the ESC script to warn the user and simply ignore it?

I'm neck-deep in macOS shit for what I expect to be the better part of the day, but I could try just disabling and enabling around the `run_event(":load")` call.

Autosaves must also be disabled when input is disabled for the reasons explained above.

PS.
I tried adding arguments to `run_event()` but it was horrible, because you don't necessarily know what kind of event is coming in. Maybe something would need to be expanded in the ESC language, like to disable input on a "ready" would be `:ready false` but that's magic and doesn't work with events like `:use i/inv_foo` without extra parsing work.

**Comment by fleskesvor** [comment]
As mentioned on IRC, I think at least `catch_input` is descriptive of what it currently does, which is to catch all mouse input to prevent the other nodes in the scene tree from receiving input when telon is active, which is useful for things like blocking input while switching between rooms or preventing the player from interacting with the room while dialog is in progress. `release_input` might not be the best naming, since input events aren't collected to be released, but it feels somewhat consistent with its opposite, even though `ignore_input` might be a more fitting name.

**Comment by mjtorn** [comment]
I'm sorry I didn't take notes last week when I worked on this, but I got all sorts of failure states, like `:ready` being blocked from input because it's an event, some kind of race condition releasing the "lock" too soon if trying to tell telon whether or not to block input, the (0, 0) coordinates etc.

This is why I hope someone who has time and understanding could take a second look, because I couldn't get them unified without one breaking the other in some way.
